backend/src/modules/authentication/core/domain/UserAccount.ts

```diff
@@ -34,9 +41,37 @@ export function setPasswordForUserAccount(
   };
 }
 
-function onPasswordWasSet(state: UserAccount | undefined, event: PasswordWasSet): UserAccount {
+function onPasswordWasSet(state: UserAccount, event: PasswordWasSet): UserAccount {
   return new UserAccount({
-    email: event.email,
+    userId: event.userId,
+    email: state.email,
     password: event.password,
   });
 }
+
+export async function authenticateUser(
+  state: UserAccount | undefined,
+  command: { email: string; password: string },
+  currentTime: Date,
+): Promise<DomainCommandResult<string | undefined>> {
+  if (!state) {
+    throw new Error('Such email address does not exists.');
+  }
+
+  const isPasswordCorrect = await bcrypt.compare(command.password, state.password!);
+  if (!isPasswordCorrect) {
+    throw new Error('Wrong password.');
+  }
+
+  const token: string = jwt.sign({ email: command.email, userId: state.userId }, `${process.env.JWT_SECRET_KEY}`, { expiresIn: '1h' });
```

Comment:
Tutaj mamy 2 rzeczy, ktore powinny byc w warstwie infrastruktury - hashowanie / porownywanie hasla i jwt. 
Przydalyby sie interfejsy Np. TokenGenerator i jakis PasswordEncpryptor. 

Bo teraz nawet wlasnie nie masz jak przetestowac tez. A w tescie np. zamockujesz, ze taki PasswordEncryptor zwrocil np. ze hasla sie nie zgadzaja. I jestes w stanie podstawic cos pod jwt. A tak to tez jest losowe.

Reply:
Domena nie powinna zalezec od takich rzeczy, jak i odczytywac w sobie process.env. To takie pure function powinny byc: 

https://enterprisecraftsmanship.com/posts/what-is-functional-programming/

https://enterprisecraftsmanship.com/posts/domain-model-purity-current-time/

https://enterprisecraftsmanship.com/posts/domain-model-purity-completeness/

Reply:
Dlatwego zapewne nie ma testów na ten Command :)

Reply:
@nowakprojects to miałeś na myśli? 
jeśli tak, to w testach, tak jak napisałeś, zamockować wyniki tych metod z interfejsów i posprawdzać różne warunki i będzie ok, tak?
https://github.com/nowakprojects/CodersCamp2020.Project.FullStack-Node-React.TableSoccerTournaments/pull/211/commits/f327b9e20a91e7e19c7bee7b06d4f93be027e877

Reply:
+ jakieś masz może pomysł na jakieś lepsze nazwy interfejsów :-P i gdzie dać ich implementacje?